internal/network/nsx_subresource.go

```diff
@@ -0,0 +1,160 @@
+/*
+ *  Copyright 2023 VMware, Inc.
+ *    SPDX-License-Identifier: MPL-2.0
+ */
+
+package network
+
+import (
+	"fmt"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/validation"
+	validation_utils "github.com/vmware/terraform-provider-vcf/internal/validation"
+	"github.com/vmware/vcf-sdk-go/models"
+)
+
+// NsxSchema this helper function extracts the NSX schema, which
+// contains the parameters required to install and configure NSX in a workload domain.
+func NsxSchema() *schema.Resource {
+	return &schema.Resource{
+		Schema: map[string]*schema.Schema{
+			"vip": {
+				Type:         schema.TypeString,
+				Required:     true,
+				Description:  "Virtual IP address which would act as proxy/alias for NSX Managers",
+				ValidateFunc: validation_utils.ValidateIPv4AddressSchema,
+			},
+			"vip_fqdn": {
+				Type:         schema.TypeString,
+				Required:     true,
+				Description:  "FQDN for VIP so that common SSL certificates can be installed across all managers",
+				ValidateFunc: validation.NoZeroValues,
+			},
+			"license_key": {
+				Type:         schema.TypeString,
+				Required:     true,
+				Description:  "NSX license value",
+				ValidateFunc: validation.NoZeroValues,
+			},
+			"form_factor": {
+				Type:         schema.TypeString,
+				Optional:     true,
+				Description:  "NSX manager form factor",
+				ValidateFunc: validation.NoZeroValues,
+			},
+			"nsx_manager_admin_password": {
+				Type:         schema.TypeString,
+				Required:     true,
+				Sensitive:    true,
+				Description:  "NSX manager admin password (basic auth and SSH)",
+				ValidateFunc: validation_utils.ValidatePassword,
+			},
+			"nsx_manager_audit_password": {
+				Type:         schema.TypeString,
+				Optional:     true,
+				Sensitive:    true,
+				Description:  "NSX manager Audit password",
+				ValidateFunc: validation_utils.ValidatePassword,
+			},
+			"nsx_manager_node": {
+				Type:        schema.TypeList,
+				Required:    true,
+				Description: "Specification details of the NSX Manager virtual machines. 3 of these are required for the first workload domain",
```

Comment:
How's this being forced to 3?

Reply:
It is not forced directly by the Terraform Resource schema, as it is quite hard to detect that this list represents the first workload domain, created in the VCF environment. However I make use of the domains validation API, that will return a failure (failing the Terraform plan) if the user supplied nsx_manager_node are less than 3 in the case of first WLD creation, while subsequently requiring only one nsx_manager_node.